Add Typescript typings

Fdebijl opened the pull request:
Fixes #61 

I took the liberty of taking the typings I used in a personal project and adapt them for distribution as bundled typings, this way TS should be supported as per #61.

Let me know if anything needs to be changed.

Fdebijl commented:
Forgot to add `index.d.ts` to `files` in package.json 😅

As a sidenote, is there a particular reason most devDependencies are outdated?
![image](https://user-images.githubusercontent.com/18357629/75080623-c9184180-550c-11ea-9a26-57e7218a2809.png)

dylanirlbeck commented:
@Fdebijl Thanks for the PR! I'll get a proper review of this ASAP.

> As a sidenote, is there a particular reason most devDependencies are outdated? It seems like microbundle 0.4.3 can't handle ts files very well while 0.11.0 can - if possible it should probably be bumped to the latest version.

Regarding this comment: I totally see your point. The former `twitter-lite` authors/maintainers didn't have a lot of time to keep this library updated, so I'm currently attempting to get caught up on outstanding issues and the library implementation itself. Many of these dependencies should probably be bumped, so I'll open an issue to look into this. Thanks for the suggestion!

peterpme commented:
Hey @Fdebijl, I just opened up #66 and after dan reviews we can use microbundle 0.11. In your case feel free to update to microbundle 0.11 :smile: it will be merged in soon

dylanirlbeck commented on index.d.ts:
```diff
@@ -0,0 +1,129 @@
+    private url: string;
+    private oauth: string;
+    private config: TwitterOptions;
+    private client: any;
```
Could we type this? I'm looking at the `oauth-1.0a` library now and they don't have types, but I think we could at least do something like this, right? Maybe when they add types later on we can come back and change this.

If you're curious: https://github.com/ddo/oauth-1.0a/blob/8c24a413ab36c7cd049d34a3d2d16996f24da0ad/oauth-1.0a.js#L9.

Note that I haven't done much work with TypeScript so please correct me if I'm wrong 😄 

```typescript
type OAuthClient = object
...
private client: OAuthClient
```

naorzr commented on index.d.ts:
```diff
@@ -0,0 +1,129 @@
+  }
+
+  class Twitter {
+    private authType: 'App' | 'User';
```
now with typescript 3.8, we can use # instead of private

naorzr commented on index.d.ts:
```diff
@@ -0,0 +1,129 @@
+import { Header } from 'oauth-1.0a';
+
+declare namespace TwitterLite {
+  type TwitterOptions = {
```
when possible interface is preferable over type
https://www.typescriptlang.org/docs/handbook/advanced-types.html#interfaces-vs-type-aliases

naorzr commented on index.d.ts:
```diff
@@ -0,0 +1,129 @@
+  }
+
+  type RequestTokenResponse = {
+    oauth_token: string;
```
oauth_token can be extracted to be its own type
and then used here and below.

Fdebijl replied:
Could you give an example? In my understanding that would just make it an alias to `string`.

dylanirlbeck replied:
@Fdebijl I think that's what he means, and I agree with the the reasoning. Any function parameter that repeats should probably be pulled out into it's own type for extensibility purposes (as another example, the `resource` parameter in `GET`, `POST`, and `PUT`). I'm not super familiar with TypeScript patterns, though, so take what I say with a grain of salt.

naorzr replied:
An example would be
``` 
type oauth_token = string;
```
and then simply using it as 
```
type RequestTokenResponse = {
    oauth_token: oauth_token;
}
@dylanirlbeck Yea that's pretty much the purpose, centralizing it and benefiting from anything that comes naturally from centralizing stuff.

Fdebijl commented:
Thank you both for your feedback! Turns out oauth-1.0a does bundle typings with their module, I've imported those rather than reverse engineering my own.

dylanirlbeck commented:
@Fdebijl Could you take a look at @naorzr's comments from your original PR? We'd like to get this merged so we could get TypeScript support 😄 Thanks again for doing this!

Fdebijl commented:
- Changed optional parameters for TwitterOptions
- Added type OauthToken (string)
- Added type OauthTokenSecret (string) 
- Added type AuthType ('App' | 'User') 
- Added interface KeySecret so that AccessTokenOptions can extend it and Twitter.#token can implement it (there was some duplication there).

peterpme commented:
Hey @Fdebijl thank you so much! Right now we're reviewing using Jared Palmer's [tsdx](https://github.com/jaredpalmer/tsdx) to make this easy moving forward :)

WIStudent commented:
I just tried this library and the typescript types for a simple twitter bot and have some remarks:

`Promise<Object>` as return type in get/post/put does fit for endpoints that return arrays, like https://developer.twitter.com/en/docs/tweets/timelines/api-reference/get-statuses-user_timeline for example.

Another huge improvement would be to overload get/post/put in a way that they only accept valid parameters and return the correct type. An example:
```typescript
interface Tweet {
  created_at: string;
  id: Number;
  id_str: string;
  text: string;
  // And so on, Sse https://developer.twitter.com/en/docs/tweets/data-dictionary/overview/tweet-object
}

interface StatusUpdateParams {
  status: string;
  in_reply_to_status_id?: string;
  auto_populate_reply_metadata?: boolean;
  exclude_reply_user_ids?: string;
  attachment_url?: string;
  media_ids?: string;
  possibly_sensitive?: boolean;
  lat?: number;
  long?: number;
  place_id?: string;
  display_coordinates?: boolean;
  trim_user?: boolean;
  enable_dmcommands?: boolean;
  fail_dmcommands?: boolean;
  card_uri?: string;
}

interface StatusesUserTimelineParams {
  user_id?: string;
  screen_name?: string;
  since_id?: string;
  count?: number;
  max_id?: string;
  trim_user?: boolean;
  exclude_replies?: boolean;
  include_rts?: boolean;
}

...
public get(resource: string, parameters?: object): Promise<Object>
public get(resource: 'statuses/user_timeline', parameters: StatusesUserTimelineParams): Promise<Tweet[]>

public post(resource: string, body: object): Promise<object>
public post(resource: 'statuses/update', body: StatusUpdateParams): Promise<Tweet>
```

By doing so someone who uses this library with typescript could directly see in their IDE which options are available/required/optional for an endpoint and what an endpoint returns.

Fdebijl commented:
@WIStudent I agree that would be very nice to have, but consider that there are [over 100 endpoints](https://developer.twitter.com/en/docs/api-reference-index) that would each need a separate overload. I fear the typings and IDE suggestions/autofill could become too crowded with that many overloads.

T0TProduction commented:
> @WIStudent I agree that would be very nice to have, but consider that there are [over 100 endpoints](https://developer.twitter.com/en/docs/api-reference-index) that would each need a separate overload. I fear the typings and IDE suggestions/autofill could become too crowded with that many overloads.

Imho an IDE would not be overloaded by this. Essentially what it would do is:
Once you choose the endpoint, you get the correct response type without needing to add anything else.
If every endpoint just returns an Object the typings aren't nearly as helpful as they could be.

The point of over 100 endpoints to me though is: We don't need all the overloads for a first TS release. This can be added later, and potentially even over time, falling back to the default return type if there is no overload defined.

dylanirlbeck commented on index.d.ts:
```diff
@@ -0,0 +1,131 @@
+     * @returns {Promise<object>} Promise resolving to the response from the Twitter API.
+     *   The `_header` property will be set to the Response headers (useful for checking rate limits)
+     */
+    public get(resource: string, parameters?: object): Promise<Object>;
```
Should the return type here be `Promisie<Object> | Promise<Array>`? Someone pointed out that some endpoints might return an array, so we should keep that in mind. If this is not a problem though we can merge, but it seems like it could be.

T0TProduction replied:
Tbh making it Promise<any> might be the best bet, because currently it *could be* anything. This also implies that its not really typed there. Only problem would be of projects don't allow any, but maybe they can fix this by immediately typecasting. IDK how any from libraries would be handled by that TS rule

T0TProduction replied:
*`Promise<any>` instead of Promise

Fdebijl replied:
I agree with @T0TProduction, going with `Promise<any>` for now seems like the best option, especially if we're going to add overloads for each endpoint. Those overloads should be the subject for a new PR though, so that people relying on typings can start working with these 'basic' ones.

naorzr replied:
I disagree with ```Promise<any>```, it would be better to just do
```public get<T>(resource: string, parameters?: object): Promise<T>;```
for the time being and let the user pass the expected result instead of casting it in his code.

T0TProduction replied:
good point. Will T default to any if not specified? not sure about that behavior. If so, I agree.

Fdebijl replied:
I reckon  ` public get<T = any>(resource: string, parameters?: object): Promise<T>;` would provide that behaviour.

naorzr replied:
Passing a generic is overall more user friendly and correct then explicit casting

Fdebijl replied:
Ammended the PR with your suggestions. What do you think @naorzr?

naorzr replied:
wouldn't 
```public get<T>(resource: string, parameters?: object): Promise<T>;```
cause the transpiler to complain in case it doesn't get a generic?
I think it's better to just test it out, if the transpiler doesn't complain then leave it as it is, if it does, you can just default it with
```public get<T = any>(resource: string, parameters?: object): Promise<T>;```
Other than that, looks good, thanks for the efforts!

Fdebijl replied:
It becomes `unknown`, which seems undesirable. I'll add `any` as a default to the type params.

Fdebijl commented:
Changed return type for the get, post and put methods to `Promise<any>`
